apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.UnitTests/TypeTests.cs

```diff
@@ -0,0 +1,627 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Google.Protobuf;
+using Google.Protobuf.WellKnownTypes;
+using Xunit;
+
+namespace Google.Cloud.Spanner.Data.Tests
+{
+    //more exhaustive (compared to integration tests) type conversion test cases.
+    public class TypeTests
+    {
+        public enum TestType
+        {
+            //Only test the CLR type -> Value conversion
+            ClrToValue = 0,
+
+            //Only test the Value -> CLR Type conversion
+            ValueToClr = 1,
+
+            //Test both ways.  There can be no loss of information or precision in the ClrToValue conversion.
+            Both = 2
+        }
+
+        private static readonly DateTime s_testDate = new DateTime(2017, 1, 31, 3, 15, 30, 500);
+        private static readonly byte[] s_bytesToEncode = {1, 2, 3, 4};
+        private static readonly string s_base64Encoded = Convert.ToBase64String(s_bytesToEncode);
+
+        private static readonly SpannerDbType s_struct = SpannerDbType.StructOf(
+            new Dictionary<string, SpannerDbType>
+            {
+                {"StringField", SpannerDbType.String},
+                {"Int64Field", SpannerDbType.Int64},
+                {"Float64Field", SpannerDbType.Float64},
+                {"BoolField", SpannerDbType.Bool},
+                {"DateField", SpannerDbType.Date},
+                {"TimestampField", SpannerDbType.Timestamp}
+            });
+
+        //nested complex type support.
+        private static readonly SpannerDbType s_arrayOfStruct = SpannerDbType.ArrayOf(s_struct);
+
+        private static readonly SpannerDbType s_complexStruct = SpannerDbType.StructOf(
+            new Dictionary<string, SpannerDbType>
+            {
+                {"StructField", s_struct},
+                {"ArrayField", SpannerDbType.ArrayOf(SpannerDbType.Int64)}
+            });
+
+        private static string S(string s) => $"\"{s}\"";
+
+        private static IEnumerable<string> GetStringsForArray()
```

Comment:
i wanted some of the test cases to be a raw enumerable backed by a method to ensure that also worked.